blocks/bricks/bn.py

```diff
@@ -0,0 +1,345 @@
+import collections
+
+import numpy
+from picklable_itertools.extras import equizip
+import theano
+from theano import tensor
+from theano.tensor.nnet import bn
+
+from ..graph import add_annotation
+from ..initialization import Constant
+from ..roles import (WEIGHT, BIAS, BATCH_NORM_POPULATION_MEAN,
+                     BATCH_NORM_POPULATION_STDEV, BATCH_NORM_OFFSET,
+                     BATCH_NORM_DIVISOR, BATCH_NORM_MINIBATCH_ESTIMATE,
+                     add_role)
+from ..utils import (shared_floatx_zeros, shared_floatx,
+                     shared_floatx_nans)
+from .base import lazy, application
+from .sequences import Sequence, Feedforward, MLP
+from .interfaces import RNGMixin
+
+
+def _add_batch_axis(var, name=None):
+    """Prepend a singleton axis to a TensorVariable."""
+    new_var = var.dimshuffle('x', *list(range(var.ndim)))
+    new_var.name = name
+    return new_var
+
+
+def _add_role_and_annotate(var, role, annotations=()):
+    """Add a role and zero or more annotations to a variable."""
+    add_role(var, role)
+    for annotation in annotations:
+        add_annotation(var, annotation)
+
+
+class BatchNormalization(RNGMixin, Feedforward):
+    r"""Normalizes activations, parameterizes a scale and shift.
+
+    Parameters
+    ----------
+    input_dim : int or tuple
+        Shape of a single input example. It is assumed that a batch axis
+        will be prepended to this.
+    broadcastable : tuple, optional
+        Tuple the same length as `input_dim` which specifies which of the
+        per-example axes should be averaged over to compute means and
+        standard deviations. For example, in order to normalize over all
+        spatial locations in a `(batch_index, channels, height, width)`
+        image, pass `(False, True, True)`.
+    save_memory : bool, optional
```

Comment:
I don't like the use of "save" here as it could be taken to mean something is being saved to somewhere. Not sure I have a better suggestion though. "Conserve" maybe?